vllm/attention/layer.py

```diff
@@ -214,20 +217,30 @@ def forward(
                                   self_kv_cache,
                                   attn_metadata,
                                   output=output)
+                maybe_save_kv_layer_to_connector(self.layer_name,
+                                                 self.kv_cache)
             else:
                 torch.ops.vllm.unified_attention_with_output(
                     query, key, value, output, self.layer_name)
+                maybe_save_kv_layer_to_connector(self.layer_name,
+                                                 self.kv_cache)
             return output.view(-1, hidden_size)
         else:
             if self.use_direct_call:
                 forward_context = get_forward_context()
                 attn_metadata = forward_context.attn_metadata
                 self_kv_cache = self.kv_cache[forward_context.virtual_engine]
-                return self.impl.forward(self, query, key, value,
-                                         self_kv_cache, attn_metadata)
+                output = self.impl.forward(self, query, key, value,
+                                           self_kv_cache, attn_metadata)
+                maybe_save_kv_layer_to_connector(self.layer_name,
+                                                 self.kv_cache)
+                return output
             else:
-                return torch.ops.vllm.unified_attention(
+                output = torch.ops.vllm.unified_attention(
                     query, key, value, self.layer_name)
+                maybe_save_kv_layer_to_connector(self.layer_name,
```

Comment:
I think that having multiple calls to `maybe_save_kv_layer_to_connector` is not ideal from a readability and maintenance perspective.

Instead, I think that we should remove the early returns and have a single call at the end after the if...elses are done